insert columnar data

Anexen opened the pull request:
There is some overhead when you trying to insert columnar data (e.g. read from parquet files) into Clickhouse via clickhouse-driver because you need to transpose data for `execute` method. Transposing data is both memory and time consuming, furthermore, the native protocol was designed to send data by columns.
My simple [benchmark](https://gist.github.com/Anexen/ceaabc2346c575859ea0d71ffc2b53ef) shows ~20% decrease in both insertion time and memory consumption while inserting columnar data.

xzkostyan commented on clickhouse_driver/block.py:
```diff
@@ -32,47 +30,130 @@ def read(self, buf):
+
+
+class ColumnOrientedBlock(BaseBlock):
+    tuple_column_types = (list, tuple)
```
It seems to be redundant.

Anexen replied:
Removed

xzkostyan commented on clickhouse_driver/client.py:
```diff
@@ -177,8 +177,10 @@ def execute(self, query, params=None, with_column_types=False,
-                         column-oriented form.
+        :param columnar: if specified the result of the SELECT query will be
+                         returned in column-oriented form.
+                         It also allows to INSERT columnar data (`list` of
```
You can also insert list of dicts.

Maybe "It also allows to INSERT data in columnar form." will be enough?

Anexen replied:
I have no idea how list of dicts should look for columnar data, so I decided to stick to list of tuples.

xzkostyan commented on clickhouse_driver/block.py:
```diff
@@ -32,47 +30,130 @@ def read(self, buf):
+        raise NotImplementedError
+
+    def transposed(self):
+        return list(map(tuple, zip(*self.data)))
```
Very good! This also faster than chaining into one big tuple and slicing.

In [the very beginning](https://github.com/mymarilyn/clickhouse-driver/blob/5d5f889c2fad2ead0d7edcf47754400a2d2b747b/src/block.py#L45) it was there and later was re-writed due to performance issues. It seems that issues were in another place.

Anexen commented on tests/test_insert.py:
```diff
@@ -147,3 +147,57 @@ def test_insert_return(self):
+                    types_check=True, columnar=True
+                )
+
+            self.assertIn('list or tuple is expected', str(e.exception))
```
As for now, the only valid format for columnar data is `list` of `tuples` where each `tuple` represents all values of particular column, e.g. `[(1,2,3), ('a', 'b', 'c')]`. I just wanted to show that all list items will be checked, not only the first one.

I think the valid `dict` format for columnar data will be something like this:
```
{
    'a': [1,2,3],
    'b': ['a', 'b', 'c']
}
```
I can support this format as well if you found it useful. It would be great if you have a suggestion about what's the best way to distinguish insert data from query params.

Anexen commented on clickhouse_driver/util/helpers.py:
```diff
@@ -7,3 +7,22 @@ def chunks(seq, n):
+        if not isinstance(column, (list, tuple)):
+            raise TypeError(
+                'Unsupported column type: {}. list or tuple is expected.'
+                .format(type(columns))
```
Yes, fixed

Anexen commented on tests/test_insert.py:
```diff
@@ -147,3 +147,57 @@ def test_insert_return(self):
+                data = [1]
+                self.client.execute(
+                    'INSERT INTO test (a) VALUES', data, columnar=True
+                )
```
I've removed this test in favor of `InsertColumnarTestCase.test_data_invalid_types`

Anexen commented:
Seems like the latest lz4 release (3.0.2) doesn't work with PyPy 3.5 (3.0.1 works as expected)

Anexen commented:
@xzkostyan, I've updated PR, please take a look. 
I've updated lz4 requirement because the latest release breaks pypy tests.

In this PR, I decided to stick to list of tuples/lists as columnar data format. 
Other options which could be implemented:
1. dict of lists
```
data = {
    'a': [1, 2, 3],
    'b': ['a', 'b', 'c']
}
```
but I think it's pretty simple to convert such dict into list of lists using the following code: `list(data.values())`  or even `data.values()` (to ensure ordering: `[data[c] for c in columns]`)

2. list of dicts `[{'a': [1, 2, 3]}, {'b':['a', 'b', 'c']}]`. But it's hard to me to imagine the use cases for this format.
3. list of "items": `[('a', [1, 2, 3]), ('b': ['a', 'b', 'c']))]`

Please let me know if you found some of this options useful.

xzkostyan commented:
@anexen I think currently other columnar formats are rather exotic than useful.
